packages/daemon/DESIGN.md

```diff
@@ -0,0 +1,215 @@
+unnecessary round trips through the underlying CapTP connection.
+
+[^pipeline]: See [Mark's explanation of pipelines][pipeline-erights], the
+    [Cap'n Proto explanation][pipeline-capnproto], [Wikipedia][pipeline-wikipedia], or
+    the [ECMAScript standard proposal][pipeline-proposal].
+
+[pipeline-erights]: http://erights.org/elib/distrib/pipeline.html
+[pipeline-capnproto]: https://capnproto.org/rpc.html
+[pipeline-wikipedia]: https://en.wikipedia.org/wiki/Futures_and_promises#Promise_pipelining
+[pipeline-proposal]: https://www.proposals.es/proposals/Support%20for%20Distributed%20Promise%20Pipelining
+
+Much hinges on what happens when a program restarts.
+Like most processes, Endo workers start without any memory from their previous
+run.
+So, Endo programs must manually persist[^persistence] anything they wish to
+remember after a restart.
+
+[^persistence]: Aside: Manual persistence is as opposed to Orthogonal
+    Persistence, where a program resumes exactly where it left off when it
+    restarts, albeit it by replaying every message it recieved since the last
+    snapshot or inception.  The architecture of Endo allows for Orthogonal
+    Persistence to be built on top of a manually persisted worker and a durable
+    reference to external storage for snapshots, message log, and virtual
+    objects.
+
+Endo applications ultimately receive all their data and powers from the Endo
+daemon with the user's consent.
+Endo persists **data** between restarts with a content address store in the
+user's home directory.
+Endo persists **powers** between restarts by storing formulas in the user's
+home directory.
+A formula describes how Endo should recreate a value, including powerful
+capability bearing objects if it restarts.
+
+When Endo runs a confined program, it provides a Powers object that the
+confined program can use to **request** powers from the user.
+The Powers object is entangled with a corresponding Inbox.
+The user interacts with the Inbox to respond to requests from the Powers
+holder.
+
+Both the Inbox held by the user (host) and the Powers held by program (guest)
+refer to formulas with pet names, and have separate pet name spaces.
+Changing a pet name does not alter or delete the underlying formula, but a
+formula that is unreachable from the user's pet names may be garbage collected.
+
+- The program can use the Powers object to request an ephemeral value
+  from the user.
+  The Powers will not retain the resulting formula.
+
+  ```js
+  const file = E(powers).request('a file');
+  ```
+
+- The program can use the Powers object to request a durable value
+  and give it a pet name.
+  The Powers will memoize the value and retain the resulting formula.
+  Endo will honor subsequent requests for the same pet name from this Powers
+  object by reconstructing the value from its formula if necessary.
+
+  ```js
+  const file = E(powers).request('a file, 'fileName');
```

Comment:
```suggestion
  const file = E(powers).request('a file', 'fileName');
```

Reply:
I am still confused that the program has a petname namespace that is distinct from the user's petname namespace.

Reply:
In the Pet Daemon spike, I’ve just made the PetStore type explicit https://github.com/endojs/endo/pull/1635/files#diff-89066b5795eb4c834b09314c6f49442ccfbfd6ea1fbc91c757cca1a6a29c88a6R117-R122

I’m open to giving this a more generic name like FormulaNameStore or Apothecary since only one instance of it is what you would properly call the user’s “pet name store” or just “pet store”.

The Endo Daemon arranges for every guest program (application, extension, caplet) to have its own “formula name store” so that:
- If the Endo Daemon restarts, the guest program (application, extension, caplet) can reconnect to powers the user had previously granted. The pet store tracks (and retains!) the formula tree needed by the guest.
- Formula garbage collection: the Endo Daemon can collect any formula that’s not reachable started at the user’s pet store, through any of the formulas transitively retained, and for any _pet store formula_, any of the formula’s retained by that pet store.